### PR TITLE
[Inference] Kernel: no pad rotary embedding

### PR DESCRIPTION
**Pull request by CjhHa1**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`
continues  #5193 

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.
Add no pad rotary embedding kernel. It should be applied to tensors after `linear projection`, followed by `memcpy`

CO-Contributor: @Xu-Kai 

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

image

### CODE REVIEW

**Review by FrankLeeeee — APPROVED** [review]

File: tests/test_infer_ops/triton/test_rotary_embdding_unpad.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    cos_shape = (TOTAL_TOKENS, D // 2)
+    cos = -1.2 + 0.5 * torch.randn(cos_shape, dtype=dtype, device="cuda")
+    sin = -2.0 + 0.5 * torch.randn(cos_shape, dtype=dtype, device="cuda")
+
+    q_ref = torch_rotary_emb(q, cos, sin)
+    k_ref = torch_rotary_emb(k, cos, sin)
+    rotary_embedding(q, k, cos, sin)
+
+    assert torch.allclose(q, q_ref, atol=1e-2, rtol=1e-3)
+    assert torch.allclose(k, k_ref, atol=1e-2, rtol=1e-3)
```
Comment (FrankLeeeee) [inline tests/test_infer_ops/triton/test_rotary_embdding_unpad.py:52]: are atol and rtol  too large?

Comment (FrankLeeeee) [reply tests/test_infer_ops/triton/test_rotary_embdding_unpad.py:52]: have you tried smaller values?

Comment (CjhHa1) [reply tests/test_infer_ops/triton/test_rotary_embdding_unpad.py:52]: tried `atol=1e-3`, can't all clone

Comment (FrankLeeeee) [reply tests/test_infer_ops/triton/test_rotary_embdding_unpad.py:52]: ok

Comment (CjhHa1) [reply tests/test_infer_ops/triton/test_rotary_embdding_unpad.py:52]: tried `atol = 1e-4` and dtype = fp32, seems like it depends on the floats

Comment (CjhHa1) [reply tests/test_infer_ops/triton/test_rotary_embdding_unpad.py:52]: image
changed already

**Pull request merged**